src/caffe/layers/spp_layer.cpp

```diff
@@ -0,0 +1,137 @@
+#include <algorithm>
+#include <cfloat>
+#include <vector>
+
+#include "caffe/common.hpp"
+#include "caffe/layer.hpp"
+#include "caffe/syncedmem.hpp"
+#include "caffe/util/math_functions.hpp"
+#include "caffe/vision_layers.hpp"
+
+namespace caffe {
+
+using std::min;
+using std::max;
+
+template <typename Dtype>
+void SPPLayer<Dtype>::LayerSetUp(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+  SPPParameter spp_param =
+      this->layer_param_.spp_param();
+
+  bottom_h_ = bottom[0]->height();
+  bottom_w_ = bottom[0]->width();
+  CHECK_GT(bottom_h_, 0) << "Input dimensions cannot be zero.";
+  CHECK_GT(bottom_w_, 0) << "Input dimensions cannot be zero.";
+
+  pyramid_height_ = spp_param.pyramid_height();
+  pad_h_ = spp_param.pad_h();
+  pad_w_ = spp_param.pad_w();
+  pooling_layers_.clear();
+  pooling_top_vecs_.clear();
+  pooling_outputs_.clear();
+  flatten_layers_.clear();
+  flatten_top_vecs_.clear();
+  flatten_outputs_.clear();
+  concat_bottom_vec_.clear();
+
+  for (int i = 0; i < pyramid_height_; i++) {
+    // pooling layer output holders setup
+    pooling_outputs_.push_back(new Blob<Dtype>());
+    pooling_top_vecs_.push_back(new vector<Blob<Dtype>*>);
+    pooling_top_vecs_[i]->push_back(pooling_outputs_[i]);
+
+    // pooling layer setup
```

Comment:
I don't think there is a way to change parameters without deleting and reinitializing them -- you could add a setter to `Layer` but I don't think it would really save anything since the constructor itself is probably basically free (`SetUp` is probably a little more expensive but you'd have to call that regardless).  Do you know if it's an issue in practice?